services/actions/cleanup.go

```diff
@@ -123,3 +129,20 @@ func CleanupLogs(ctx context.Context) error {
 	log.Info("Removed %d logs", count)
 	return nil
 }
+
+const deleteEphemeralRunnerBatchSize = 100
+
+// CleanupEphemeralRunners removes used ephemeral runners which are no longer able to process jobs
+func CleanupEphemeralRunners(ctx context.Context) error {
+	runners := []*actions_model.ActionRunner{}
+	err := db.GetEngine(ctx).Join("INNER", "`action_task`", "`action_task`.`runner_id` = `action_runner`.`id`").Where("`action_runner`.`ephemeral` and `action_task`.`status` != ? and `action_task`.`status` != ? and `action_task`.`status` != ?)", actions_model.StatusWaiting, actions_model.StatusRunning, actions_model.StatusBlocked).Limit(deleteEphemeralRunnerBatchSize).Find(&runners)
+	if err != nil {
+		return fmt.Errorf("find runners: %w", err)
+	}
+	count, err := db.GetEngine(ctx).Delete(&runners)
```

Comment:
IIRC `Delete(&runners)` here would generate quite complex SQLs since the `runners` have a lot of fields?